From: Pritha Srivastava <prsrivas@redhat.com>
Date: Mon, 29 Aug 2016 14:32:42 +0530
Subject: rgw: Replacing '+' with "%20" in canonical uri for s3 v4 auth.

s3cmd encodes space as "%20" while signature computation and
encodes space as '+' while sending the canonical uri. This
results in a SignatureMismatch Error in rgw, since rgw
computes the signature based on the request received from
the client (s3cmd in this case).

Fixes http://tracker.ceph.com/issues/17076.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>
(cherry picked from commit 20e5ff023ebad89c386a520d07613547d4836399)

X-short-title: AWS S3 Version 4 signatures fail sometimes
X-tracker-ref: http://tracker.ceph.com/issues/17076
X-github-pr: https://github.com/ceph/ceph/pull/12542
---
 src/rgw/rgw_rest_s3.cc | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/rgw/rgw_rest_s3.cc b/src/rgw/rgw_rest_s3.cc
index da93a95..fcd336a 100644
--- a/src/rgw/rgw_rest_s3.cc
+++ b/src/rgw/rgw_rest_s3.cc
@@ -10,6 +10,7 @@
 #include "common/ceph_json.h"
 #include "common/safe_io.h"
 #include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/replace.hpp>
 
 #include "rgw_rest.h"
 #include "rgw_rest_s3.h"
@@ -3580,6 +3581,8 @@ int RGW_Auth_S3::authorize_v4(RGWRados *store, struct req_state *s)
 
   if (s->aws4_auth->canonical_uri.empty()) {
     s->aws4_auth->canonical_uri = "/";
+  } else {
+    boost::replace_all(s->aws4_auth->canonical_uri, "+", "%20");
   }
 
   /* craft canonical query string */
